### PR TITLE
feat: UndoManager

### PR DESCRIPTION
**Pull request by sirskunkalot**

### CODE REVIEW

File: JotunnLib/Managers/UndoManager.cs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+            var hud = MessageHud.instance;
+            if (!hud)
+            {
+                return;
+            }
+            if (priority)
+            {
+                var items = hud.m_msgQeue.ToArray();
+                hud.m_msgQeue.Clear();
+                Player.m_localPlayer?.Message(MessageHud.MessageType.TopLeft, message);
```
Comment (MSchmoecker) [inline JotunnLib/Managers/UndoManager.cs]: code style: Player.m_localPlayer is a Unity Object and should not be conditional chained

File: JotunnLib/Managers/UndoManager.cs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                foreach (var item in items)
+                {
+                    hud.m_msgQeue.Enqueue(item);
+                }
+                hud.m_msgQueueTimer = 10f;
+            }
+            else
+            {
+                Player.m_localPlayer?.Message(MessageHud.MessageType.TopLeft, message);
```
Comment (MSchmoecker) [inline JotunnLib/Managers/UndoManager.cs]: code style: Player.m_localPlayer is a Unity Object and should not be conditional chained

File: JotunnLib/Utils/UndoActions.cs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+namespace Jotunn.Utils
+{
+    /// <summary>
+    ///     Pre-defined actions to use with the <see cref="UndoManager"/>.
+    /// </summary>
+    public static class UndoActions
+    {
+        private static class UndoHelper
+        {
+            public static void CopyData(ZDO from, ZDO to)
```
Comment (MSchmoecker) [inline JotunnLib/Utils/UndoActions.cs:16]: Maybe put `zdo.Clone()` into this function and not rely on an external call? It always have to be made and if it is forgotten, the original ZDO properties are edited

File: JotunnLib/Utils/UndoActions.cs [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+                to.m_longs = from.m_longs;
+                to.m_strings = from.m_strings;
+                to.m_byteArrays = from.m_byteArrays;
+                to.m_prefab = from.m_prefab;
+                to.m_position = from.m_position;
+                to.m_rotation = from.m_rotation;
+                var zs = ZNetScene.instance;
+                if (zs.m_instances.TryGetValue(to, out var view))
+                {
+                    view.transform.position = from.m_position;
```
Comment (MSchmoecker) [inline JotunnLib/Utils/UndoActions.cs]: code style/performance: view.transform should only be called once

File: JotunnLib/Utils/UndoActions.cs [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+                compiler.m_lastOpRadius = 0f;
+                compiler.Save();
+                compiler.m_hmap.Poke(false);
+            }
+        }
+        
+        /// <summary>
+        ///     "Create" action for the <see cref="UndoManager"/>. Can undo and redo ZDO creation.
+        /// </summary>
+        public class UndoCreate : UndoManager.IUndoAction
```
Comment (MSchmoecker) [inline JotunnLib/Utils/UndoActions.cs:138]: UndoCreate/Remove may be an ambiguous name, maybe UndoCreateZDOs?

File: JotunnLib/Utils/UndoActions.cs [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                        compiler.m_modifiedHeight[value.Index] = value.HeightModified;
+                    }
+                    foreach (var value in kvp.Value.Paints)
+                    {
+                        compiler.m_modifiedPaint[value.Index] = value.PaintModified;
+                        compiler.m_paintMask[value.Index] = value.Paint;
+                    }
+                    Save(compiler);
+                }
+                ClutterSystem.instance?.ResetGrass(pos, radius);
```
Comment (MSchmoecker) [inline JotunnLib/Utils/UndoActions.cs:120]: code style: ClutterSystem.instance is a Unity Object and should not be conditional chained

**Review by MSchmoecker — APPROVED** [review]

File: JotunnLib/Utils/UndoActions.cs (hunk 3 above)
Comment (sirskunkalot) [reply JotunnLib/Utils/UndoActions.cs:16]: nope, CopyData must not Clone() or it has to be DeepCopyData() ;)

**Pull request merged**